NASA-PDS/validate
# Fix bug when checking for offset for fields that are not of BIT type

1. Add test resource : src/test/resources/github392
2. Add logic to check for data_type of field before making the offset checks : src/main/java/gov/nasa/pds/tools/validate/content/table/FieldValueValidator.java
3. Add 3 new tests : src/test/resources/features/validate.feature

Refs:

https://github.com/nasa-pds/validate/issues/392 Validate throws incorrect overlap error when first Field_Bit has length 1

Closes https://github.com/nasa-pds/validate/issues/392 Validate throws incorrect overlap error when first Field_Bit has length 1

The data_type UnsignedBitString requires some refinement in checking for offset of each field.  Validate must know
what the data_type is for a particular field before making the check.  The check in question is modified to only performing the normal
offset between one field to another if the data_type is NOT UnsignedBitString.

Tested in DEV:

Case 1:
Running validate against a valid label with with data_type UnsignedBitString.  No errors expected.

```
% validate -R pds4.label --skip-context-validation -r report_github392_label_valid.json -s json -t src/test/resources/github392/test1_valid.xml
```

Case 2: 
Running validate against a label with Table_Binary with data_type UnsignedBitString with two bit fields overlap.

```
% validate -R pds4.label --skip-context-validation -r report_github392_label_invalid_1.json -s json -t src/test/resources/github392/test1_invalid.xml
```

totalErrors = 1
   error.table.field_value_overlap 1
              "message": "The bit field overlaps the next field. Current stop_bit_location: 7. Next start_bit_location: 6"

Case 3: 
Running validate against a label with Table_Binary with data_type SignedMSB4 with two errors:

 1. Field 1 and field 2 overlaps
 2. The offset of 16 greater than the size of the record which is 7 bytes.

```
validate -R pds4.label --skip-context-validation -r report_github392_label_invalid_2.json -s json -t src/test/resources/github392/INVALID_odf07155_msgr_11.xml
```

totalErrors = 2
   error.table.field_value_overlap 1
             "message": "This field overlaps the next field. Current field ends at 4. Next field starts at 3 but should be at least at 5"

   error.table.bad_field_read 1
              "message": "Error while getting field value: Field offset '16' is greater than the number of bytes in the record '7'"
